src/cargo/core/compiler/context/mod.rs

```diff
@@ -67,6 +69,27 @@ pub struct Unit<'a> {
     pub kind: Kind,
 }
 
+impl<'a> Unit<'a> {
+    pub fn buildkey(self) -> String {
+        format!(
+            "{} {} {} {:?}",
+            self.pkg, self.target, self.profile, self.kind
+        )
+    }
```

Comment:
Let's switch to @alexcrichton suggestions with hashes? It could look like this I think:

```
impl<'a> Unit<'a> {
    pub fn buildkey(&self) -> String {
        #![allow(deprecated)] // for SipHash
        use std::hash::{Hash, Hasher, SipHasher};

        let mut hasher = SipHasher::new_with_keys(0, 0);
        self.hash(&mut hasher);
        format!("{}-{}", self.pkg.name(), hasher.finish())
    }
}
```

Reply:
Agreed with @matklad on this, getting all the info into the key through the hash I think would be best here to ensure we don't forget to include anything